File: ios/Sources/DeviceInfoPlugin/DeviceInfoPlugin.swift
import Capacitor
import Foundation

@objc(DeviceInfoPlugin)
public class DeviceInfoPlugin: CAPPlugin, CAPBridgedPlugin {
    private let pluginVersion = "8.0.3"
    private let defaultIntervalMs = 1000
    private let minimumIntervalMs = 250

    public let identifier = "DeviceInfoPlugin"
    public let jsName = "DeviceInfo"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "getInfo", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "startMonitoring", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "stopMonitoring", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "isMonitoring", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "removeAllListeners", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getPluginVersion", returnType: CAPPluginReturnPromise)
    ]

    private let implementation = DeviceInfo()
    private var timer: Timer?
    private var intervalMs: Int?
    private var startedAt: Double?
    private var sampleLimit: Int?
    private var stopAt: Double?
    private var samplesEmitted = 0

    @objc func getInfo(_ call: CAPPluginCall) {
        call.resolve(implementation.getInfo())
    }

    @objc func startMonitoring(_ call: CAPPluginCall) {
        stopMonitoringInternal()

        let effectiveIntervalMs = max(call.getInt("intervalMs") ?? defaultIntervalMs, minimumIntervalMs)
        let now = currentTimestamp()
        intervalMs = effectiveIntervalMs
        startedAt = now
        samplesEmitted = 0
        sampleLimit = positiveInt(call.getInt("sampleCount"))

        if let durationMs = positiveDouble(call.getDouble("durationMs")) {
            stopAt = now + durationMs
        }

        if call.getBool("emitImmediately") != false {
            emitSample()
        }

        if startedAt != nil {
            let timer = Timer(timeInterval: Double(effectiveIntervalMs) / 1000, repeats: true) { [weak self] _ in
                self?.emitSample()
            }
            RunLoop.main.add(timer, forMode: .common)
            self.timer = timer
        }

        call.resolve([
            "monitoring": startedAt != nil,
            "intervalMs": effectiveIntervalMs,
            "startedAt": now
        ])
    }

    @objc func stopMonitoring(_ call: CAPPluginCall) {
        stopMonitoringInternal()
        call.resolve(["monitoring": false])
    }

    @objc func isMonitoring(_ call: CAPPluginCall) {
        var result: [String: Any] = [
            "monitoring": timer != nil
        ]

        if let intervalMs {
            result["intervalMs"] = intervalMs
        }
        if let startedAt {
            result["startedAt"] = startedAt
        }
        if timer != nil {
            result["samplesEmitted"] = samplesEmitted
        }

        call.resolve(result)
    }

    @objc override public func removeAllListeners(_ call: CAPPluginCall) {
        super.removeAllListeners(call)
    }

    @objc func getPluginVersion(_ call: CAPPluginCall) {
        call.resolve(["version": pluginVersion])
    }

    deinit {
        stopMonitoringInternal()
    }

    private func emitSample() {
        guard let startedAt else {
            return
        }

        var sample = implementation.getInfo()
        samplesEmitted += 1
        let timestamp = sample["timestamp"] as? Double ?? currentTimestamp()
        sample["sequence"] = samplesEmitted
        sample["startedAt"] = startedAt
        sample["elapsedMs"] = timestamp - startedAt

        notifyListeners("deviceInfoUpdate", data: sample)

        if shouldStopMonitoring() {
            stopMonitoringInternal()
        }
    }

    private func shouldStopMonitoring() -> Bool {
        if let sampleLimit, samplesEmitted >= sampleLimit {
            return true
        }
        if let stopAt, currentTimestamp() >= stopAt {
            return true
        }
        return false
    }

    private func stopMonitoringInternal() {
        timer?.invalidate()
        timer = nil
        intervalMs = nil
        startedAt = nil
        sampleLimit = nil
        stopAt = nil
        samplesEmitted = 0
    }

    private func positiveInt(_ value: Int?) -> Int? {
        guard let value, value > 0 else {
            return nil
        }
        return value
    }

    private func positiveDouble(_ value: Double?) -> Double? {
        guard let value, value > 0 else {
            return nil
        }
        return value
    }

    private func currentTimestamp() -> Double {
        return Date().timeIntervalSince1970 * 1000
    }
}

File: android/src/main/java/app/capgo/deviceinfo/DeviceInfoPlugin.java
package app.capgo.deviceinfo;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.TimeUnit;

@CapacitorPlugin(name = "DeviceInfo")
public class DeviceInfoPlugin extends Plugin {

    private static final String EVENT_DEVICE_INFO_UPDATE = "deviceInfoUpdate";
    private static final int DEFAULT_INTERVAL_MS = 1000;
    private static final int MIN_INTERVAL_MS = 250;

    private final DeviceInfo implementation = new DeviceInfo();
    private final Object monitoringLock = new Object();
    private final ExecutorService readExecutor = Executors.newSingleThreadExecutor();

    private ScheduledExecutorService executorService;
    private ScheduledFuture<?> monitoringTask;
    private int intervalMs;
    private long startedAt;
    private int samplesEmitted;
    private Integer sampleLimit;
    private Long stopAt;

    @PluginMethod
    public void getInfo(PluginCall call) {
        readExecutor.execute(() -> resolveCall(call, implementation.getInfo(getContext())));
    }

    @PluginMethod
    public void startMonitoring(PluginCall call) {
        int effectiveIntervalMs = Math.max(call.getInt("intervalMs", DEFAULT_INTERVAL_MS), MIN_INTERVAL_MS);
        Integer requestedSampleLimit = call.getInt("sampleCount", null);
        Double requestedDurationMs = call.getDouble("durationMs");
        boolean emitImmediately = call.getBoolean("emitImmediately", true);
        long now = System.currentTimeMillis();

        synchronized (monitoringLock) {
            stopMonitoringInternal();
            intervalMs = effectiveIntervalMs;
            startedAt = now;
            samplesEmitted = 0;
            sampleLimit = requestedSampleLimit != null && requestedSampleLimit > 0 ? requestedSampleLimit : null;
            stopAt = requestedDurationMs != null && requestedDurationMs > 0 ? now + requestedDurationMs.longValue() : null;
            executorService = Executors.newSingleThreadScheduledExecutor();

            if (emitImmediately) {
                executorService.execute(this::emitSample);
            }

            if (isMonitoringLocked()) {
                monitoringTask = executorService.scheduleAtFixedRate(
                    this::emitSample,
                    effectiveIntervalMs,
                    effectiveIntervalMs,
                    TimeUnit.MILLISECONDS
                );
            }
        }

        JSObject ret = new JSObject();
        ret.put("monitoring", isMonitoring());
        ret.put("intervalMs", effectiveIntervalMs);
        ret.put("startedAt", (double) now);
        call.resolve(ret);
    }

    @PluginMethod
    public void stopMonitoring(PluginCall call) {
        synchronized (monitoringLock) {
            stopMonitoringInternal();
        }

        JSObject ret = new JSObject();
        ret.put("monitoring", false);
        call.resolve(ret);
    }

    @PluginMethod
    public void isMonitoring(PluginCall call) {
        JSObject ret = new JSObject();

        synchronized (monitoringLock) {
            boolean monitoring = isMonitoringLocked();
            ret.put("monitoring", monitoring);
            if (monitoring) {
                ret.put("intervalMs", intervalMs);
                ret.put("startedAt", (double) startedAt);
                ret.put("samplesEmitted", samplesEmitted);
            }
        }

        call.resolve(ret);
    }

    @PluginMethod
    public void removeAllListeners(PluginCall call) {
        super.removeAllListeners(call);
    }

    @PluginMethod
    public void getPluginVersion(PluginCall call) {
        JSObject ret = new JSObject();
        ret.put("version", implementation.getPluginVersion());
        call.resolve(ret);
    }

    @Override
    protected void handleOnDestroy() {
        synchronized (monitoringLock) {
            stopMonitoringInternal();
        }
        readExecutor.shutdownNow();
        super.handleOnDestroy();
    }

    private void resolveCall(PluginCall call, JSObject data) {
        if (getActivity() != null) {
            getActivity().runOnUiThread(() -> call.resolve(data));
        } else {
            call.resolve(data);
        }
    }

    private void emitSample() {
        JSObject sample;
        boolean shouldStop;

        synchronized (monitoringLock) {
            if (!isMonitoringLocked()) {
                return;
            }

            sample = implementation.getInfo(getContext());
            samplesEmitted += 1;
            sample.put("sequence", samplesEmitted);
            sample.put("startedAt", (double) startedAt);
            sample.put("elapsedMs", System.currentTimeMillis() - startedAt);
            shouldStop = shouldStopMonitoringLocked();
        }

        if (getActivity() != null) {
            getActivity().runOnUiThread(() -> notifyListeners(EVENT_DEVICE_INFO_UPDATE, sample));
        } else {
            notifyListeners(EVENT_DEVICE_INFO_UPDATE, sample);
        }

        if (shouldStop) {
            synchronized (monitoringLock) {
                stopMonitoringInternal();
            }
        }
    }

    private boolean shouldStopMonitoringLocked() {
        if (sampleLimit != null && samplesEmitted >= sampleLimit) {
            return true;
        }
        return stopAt != null && System.currentTimeMillis() >= stopAt;
    }

    private boolean isMonitoring() {
        synchronized (monitoringLock) {
            return isMonitoringLocked();
        }
    }

    private boolean isMonitoringLocked() {
        return executorService != null && !executorService.isShutdown() && startedAt > 0;
    }

    private void stopMonitoringInternal() {
        if (monitoringTask != null) {
            monitoringTask.cancel(false);
            monitoringTask = null;
        }
        if (executorService != null) {
            executorService.shutdownNow();
            executorService = null;
        }
        intervalMs = 0;
        startedAt = 0;
        samplesEmitted = 0;
        sampleLimit = null;
        stopAt = null;
    }
}
